openstates/utils/instrument.py

```diff
@@ -0,0 +1,168 @@
+from ast import literal_eval
+import logging
+import os
+import jwt
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+import time
+from typing import List, Dict
+
+
+class MetricTypes:
+    CounterType = "count"
+    GaugeType = "gauge"
+    TimingType = "timing"
+    SetType = "set"
+
+
+class Instrumentation(object):
+    def __init__(self) -> None:
+        """
+        We're currently leveraging https://github.com/civic-eagle/statsd-http-proxy
+        to have an easy to reason about stats aggregation tool
+        on the other end of an internet connection
+
+        Essentially statsd-style metrics, we can rely on the aggregator
+        to continually emit these stats so we get useful stats for monitoring/reporting
+        """
+        self.logger = logging.getLogger("openstates")
+        # use a literal_eval to properly turn a string into a bool (literal_eval 'cause it's safer than stdlib eval)
+        self.enabled = literal_eval(os.environ.get("STATS_ENABLED", "False"))
+        token: str = self._jwt_token()
+        self._batch: List[Dict] = list()
+        self.prefix: str = os.environ.get("STATS_PREFIX", "openstates_")
+        self.endpoint: str = os.environ.get("STATS_ENDPOINT", "")
+        stats_retries: int = int(os.environ.get("STATS_RETRIES", 3))
+        self.batch_size: int = int(os.environ.get("STATS_BATCH_SIZE", 50))
+        self.default_tags: Dict = dict()
+        headers = {"Content-Type": "application/json"}
+        if token:
+            headers["X-JWT-Token"] = token
+        retry = Retry(
+            total=stats_retries,
+            read=stats_retries,
+            connect=stats_retries,
+            backoff_factor=0.3,
+            status_forcelist=(500, 502, 503, 504, 429, 104),
+        )
+        adapter = HTTPAdapter(max_retries=retry)
+        self._stat_client = requests.Session()
+        self._stat_client.headers.update(headers)
+        # only mount https endpoint if we need it
+        if self.endpoint.startswith("https"):
+            self._stat_client.mount("https://", adapter)
+        else:
+            self._stat_client.mount("http://", adapter)
+
+    def _jwt_token(self) -> str:
+        """
+        Only generate a token at startup. Much faster.
+        """
+        if not self.enabled:
+            return ""
+        secret = os.environ.get("STATS_JWT_SECRET", "")
+        if not secret:
+            return ""
+        return jwt.encode({"id": "openstates"}, secret, algorithm="HS256")
+
+    def _send_stats(self, force: bool = False) -> None:
+        """
+        Needs to be broken out from _process_metric to have a
+        "write at shutdown" function (force=True)
+        """
+        batch_len = len(self._batch)
+        if (force and batch_len > 0) or batch_len > self.batch_size:
+            if not self.endpoint:
```

Comment:
Am I understanding this correctly: if a 3rd party uses the openstates-core CLI, and doesn't have stats configured, the CLI commands will work normally. Except they will also emit a warning for every attempt to send stats (which could be what, like a handful of them? or dozens of them?). Is that right?

If so, seems like that meets a criteria I have (folks can still use the openstates CLI without having to have the full Civic Eagle infra in place. 

It does seem like maybe one "stats are not enaled) warning at the start or end of command execution might be nicer than many of them.

Reply:
Good point. Changed the logging to only output at debug level when configs aren't set. Should make normal runs much cleaner for people.